File: ios/Classes/BarcodeHandler.swift
//
//  BarcodeHandler.swift
//  mobile_scanner
//
//  Created by Julian Steenbakker on 24/08/2022.
//

import Foundation

public class BarcodeHandler: NSObject, FlutterStreamHandler {
    
    var event: [String: Any?] = [:]
    
    private var eventSink: FlutterEventSink?
    private let eventChannel: FlutterEventChannel
    
    init(registrar: FlutterPluginRegistrar) {
        eventChannel = FlutterEventChannel(name:
                                            "dev.steenbakker.mobile_scanner/scanner/event", binaryMessenger: registrar.messenger())
        super.init()
        eventChannel.setStreamHandler(self)
    }
    
    func publishEvent(_ event: [String: Any?]) {
        self.event = event
        eventSink?(event)
    }
    
    public func onListen(withArguments arguments: Any?,
                         eventSink: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = eventSink
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}

File: android/src/main/kotlin/dev/steenbakker/mobile_scanner/BarcodeHandler.kt
package dev.steenbakker.mobile_scanner

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel

class BarcodeHandler(binaryMessenger: BinaryMessenger) : EventChannel.StreamHandler {

    private var eventSink: EventChannel.EventSink? = null

    private val eventChannel = EventChannel(
        binaryMessenger,
        "dev.steenbakker.mobile_scanner/scanner/event"
    )

    init {
        eventChannel.setStreamHandler(this)
    }

    fun publishError(errorCode: String, errorMessage: String, errorDetails: Any?) {
        Handler(Looper.getMainLooper()).post {
            eventSink?.error(errorCode, errorMessage, errorDetails)
        }
    }

    fun publishEvent(event: Map<String, Any>) {
        Handler(Looper.getMainLooper()).post {
            eventSink?.success(event)
        }
    }

    fun dispose() {
        eventChannel.setStreamHandler(null)
    }

    override fun onListen(event: Any?, eventSink: EventChannel.EventSink?) {
        this.eventSink = eventSink
    }

    override fun onCancel(event: Any?) {
        this.eventSink = null
    }
}